plush/map.pls

```diff
@@ -0,0 +1,277 @@
+    self._capacity = newCapacity;
+    self._size = 0;
+    self:_resizeArrToCapacity();
+    for (var i = 0; i < oldCapacity; i += 1)
+    {
+        var bucket = oldArr[i];
+        var len = bucket.length;
+        for (var j = 0; j < len; j += 1)
+        {
+            var entry = bucket[j];
+            self:set(entry[0], entry[1]);
+        }
+    }
+};
+
+
+/// Returns the number of entries in hashmap
+HashMap.size = function(self)
+{
+    return self._size;
+};
+
+
+/// Inserts key and value in hashmap if key is not present. Otherwise updates
+/// the value of the key with then provided value.
+HashMap.set = function(self, key, value)
+{
+    var index = self._hash(key, self._capacity);
+    var bucket = self._arr[index];
+    var len = bucket.length;
+    if (len > 0)
+    {
+        // If key exists, replace and return
+        for (var i = 0; i < len; i += 1)
+        {
+            if (bucket[i][0] == key)
+            {
+                bucket[i][1] = value;
+                return ;
+            }
+        }
+    }
+    bucket:push([key, value]);
+    self._size += 1;
+    var load = (self._size / self._capacity);
+    if (load >= maxloadFactor)
+    {
+        self:_resize(self._capacity * 2);
+    }
+};
+
+
+/// Returns value corresponding to the key if present. Otherwise throws
+/// KeyNotFound exception. Either handle the exception properly or make sure
+/// key is present before calling this function.
+HashMap.get = function(self, key)
+{
+    var index = self._hash(key, self._capacity);
+    var bucket = self._arr[index];
+    var len = bucket.length;
+    for (var i = 0; i < len; i += 1)
+    {
+        var entry = bucket[i];
+        if (entry[0] == key)
+        {
+            return entry[1];
+        }
+    }
+    throw strings.format("KeyNotFound: {}", [key]);
+};
+
+
+/// Returns value corresponding to the key if present. Otherwise
+/// returns provided value
+HashMap.getOrDefault = function(self, key, defaultValue)
+{
+
+    try
+    {
+        return HashMap.get(self, key);
+    }
+    catch (e)
+    {
+        return defaultValue;
+    }
+};
+
+
+/// Returns true if the key is present in hashmap otherwise false
+HashMap.has = function(self, key)
+{
+    var index = self._hash(key, self._capacity);
+    var bucket = self._arr[index];
+    var len = bucket.length;
+    for (var i = 0; i < len; i += 1)
+    {
+        var entry = bucket[i];
+        if (entry[0] == key)
+        {
+            return true;
+        }
+    }
+    return false;
+};
+
+
+/// Removes the entry corresponding to the key if present. Otherwise does nothing.
+HashMap.remove = function(self, key)
+{
+    var index = self._hash(key, self._capacity);
+    var bucket = self._arr[index];
+    var len = bucket.length;
+    for (var i = 0; i < len; i += 1)
+    {
+        var entry = bucket[i];
+        if (entry[0] == key)
+        {
+            var left = bucket:slice(0, i);
+            var right = bucket:slice(i+1, len);
```

Comment:
I'm a total newbie, thus what I'm going to say may be very wrong.
If we used a double linked list instead of an array to represent the bucket deletion should be quicker (less copying). Are there any reason that has led you to use an array?

Reply:
In Most of the cases array will outperfom linkedlists. Primary reason for this is that arrays are contiguous and hence cache friendly. Specially in this case when size is very small. I ran an experiment, on 10000 random string keys max chain length was 5.

Reply:
Orginally my idea was to convert the bucket to red-black tree if its size exceeds some threshold (`java.util.HashMap` does this). I didn't have enough time to do it.
In current implementation, a carefully crafted attack can degrade the preformace of most of the methods to linear. Python prevents this by having more that one hash function, one of them is chosen randomly when vm boots.
There are lot of thing to still to be done. This is just initial implementation.
Also I found reference to the famous [John bentley and Bjarne Stroustrup Experiment](https://isocpp.org/blog/2014/06/stroustrup-lists)

Reply:
Makes sense! Thanks for being patient enough to explain this! :smile: